src/os_auth/auth.c

```diff
@@ -202,24 +209,55 @@ w_err_t w_auth_replace_agent(keyentry *key,
         return OS_INVALID;
     }
 
+    if (j_agent_info = wdb_get_agent_info(atoi(key->id), NULL), !j_agent_info) {
+        mdebug1("Failed to get agent '%s' information from Wazuh DB.", key->id);
+        return OS_INVALID;
+    }
+
+    j_date_add = cJSON_GetObjectItem(j_agent_info->child, "date_add");
+    j_connection_status = cJSON_GetObjectItem(j_agent_info->child, "connection_status");
+    j_disconnected_time = cJSON_GetObjectItem(j_agent_info->child, "disconnected_time");
+
     /* Check if the agent has been disconnected longer than the value specified in the configuration option*/
     if (force_options->disconnected_time_enabled) {
-        time_t disconnected_time = 0;
-        disconnected_time = (long)get_time_since_agent_disconnection(key->id);
-        if (!disconnected_time) {
-            minfo("Agent '%s' can't be replaced since it is not disconnected.", key->id);
+        if (!cJSON_IsString(j_connection_status)) {
+            mdebug1("Failed to get connection status for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
             return OS_INVALID;
-        } else if (disconnected_time > 0 && disconnected_time < force_options->disconnected_time) {
-            minfo("Agent '%s' has not been disconnected long enough to be replaced.", key->id);
+        }
+        if (!cJSON_IsNumber(j_disconnected_time)) {
+            mdebug1("Failed to get disconnected time for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
```

Comment:
We should avoid all this duplicated cJSON_Delete(j_agent_info);
I suggest you declaring a boolean **replace_agent** initialized to true. And you set it to false on each return **OS_INVALID**, at the end of the function you check this value to eliminate or not the old agent.

Reply:
Changed